### DISCUSSION TITLE
PHP Fatal error:  Uncaught Error: Undefined constant "YOURLS_USER"

### DISCUSSION THREAD

**Discussion by dlpartain** [YOURLS/YOURLS#3703] [Q&A] [answered]
Greetings.  I have what I believe is a pretty standard 1.8.2 installation on Ubuntu 22.04.  I'm not the primary person supporting this (he's on vacation...) and am floundering a bit.  When people try to create a shortened link, the system is throwing:

PHP Fatal error:  Uncaught Error: Undefined constant "YOURLS_USER" in /var/www/html/user/plugins/yourls-domainlimit-plugin-master/plugin.php:28
Stack trace:
#0 /var/www/html/includes/functions-plugins.php(229): domainlimit_link_filter()
#1 /var/www/html/includes/functions-shorturls.php(30): yourls_apply_filter()
#2 /var/www/html/index.php(84): yourls_add_new_link()
#3 {main}
thrown in /var/www/html/user/plugins/yourls-domainlimit-plugin-master/plugin.php on line 28

I'm not well-versed at PHP, so I hope I'm not missing something sillily-obvious :)  Can anyone point me in the right direction for a solution?  Thanks!

**Comment by dgw** [comment]
You're using a very old plugin, last updated 11 years ago. That said, I'm guessing that your install has public shortening enabled, and that the error described above only happens for users who aren't signed in. YOURLS only defines `YOURLS_USER` after verifying authentication:

https://github.com/YOURLS/YOURLS/blob/b493c47e6f02a7655d388b271b042e1869bfc3e8/includes/functions-auth.php#L519-L522

If the user isn't authenticated, that constant doesn't exist. The domainlimit plugin seems to assume that only authenticated users can shorten links, and doesn't check if `YOURLS_USER` is defined before [trying to use it](https://github.com/nicwaller/yourls-domainlimit-plugin/blob/af63f31b8e82ad496364e639c820032f3b9aa20f/plugin.php#L26-L30).

This "worked" in older versions of PHP (and [became a warning in PHP 7.2](https://www.php.net/manual/en/migration72.deprecated.php#migration72.deprecated.unquoted-strings)) because an undefined constant used to be interpreted as an unquoted string, i.e. the plugin ended up checking if `'YOURLS_USER'` was an exempt username (which it wouldn't be) and just moved on with its life. [PHP 8.0 just raises an Error for this case.](https://www.php.net/manual/en/migration80.incompatible.php#:~:text=Attempting%20to%20access%20unqualified%20constants%20which%20are%20undefined.%20Previously%2C%20unqualified%20constant%20accesses%20resulted%20in%20a%20warning%20and%20were%20interpreted%20as%20strings) (Ctrl-F for "unqualified constant" if your browser doesn't support linking to highlights).

~~You can use the patched `plugin.php` file from https://github.com/nicwaller/yourls-domainlimit-plugin/pull/9 even though it's highly unlikely ever to be merged by the plugin's maintainer.~~ @nicwaller updated the plugin 🎉 and posted his own answer which supersedes mine: https://github.com/orgs/YOURLS/discussions/3703#discussioncomment-8563116

**Comment by nicwaller** [comment, marked as answer]
Upgrading the `yourls-domainlimit-plugin` plugin to version `1.1.1` should fix the problem.

---

Update this file on your server:

`/var/www/html/user/plugins/yourls-domainlimit-plugin-master/plugin.php`

To use the updated code from here:

https://github.com/nicwaller/yourls-domainlimit-plugin/blob/1.1.1/plugin.php
